kmod/src/kernelcompat.h

```diff
@@ -181,4 +181,35 @@ do {					\
 #define kc_bio_get_errno(bio)			({ (int)((void)(bio), _error_arg); })
 #endif
 
+#ifndef KC_SHRINKER_SHRINK
+#define KC_DEFINE_SHRINKER(name) struct shrinker name
+#define KC_INIT_SHRINKER_FUNCS(type, name, shrinkfn, countfn, scanfn) do {	\
+	__typeof__(shrinkfn) _shrink = (shrinkfn);				\
+	_shrink->count_objects = (countfn);					\
+	_shrink->scan_objects = (scanfn);					\
+} while (0)
+#else
+#include <linux/shrinker.h>
+int kc_shrink_wrapper(struct shrinker *shrink, struct shrink_control *sc);
+struct kc_shrinker_funcs {
+	unsigned long (*count_objects)(struct shrinker *, struct shrink_control *sc);
+	unsigned long (*scan_objects)(struct shrinker *, struct shrink_control *sc);
+};
+/* using adjacent member of an unnamed struct */
+#define KC_DEFINE_SHRINKER(name) struct				\
+	{							\
+		struct kc_shrinker_funcs shrinker_funcs;	\
+		struct shrinker (name);				\
+	}
+#define KC_SHRINKER_FUNCS(shrinkfn) \
+	((void *)((long)(shrinkfn) - sizeof(struct kc_shrinker_funcs)))
```

Comment:
I think fixing this raw pointer math is pretty easy.  We already have the kc_shrinker_funcs struct.  Rename it to a _shrinker_wrapper and put the struct shrinker in there.  The definition and init macros would change a bit, and the _SHRINKER_FUNCS could turn into a SHRINKER_WRAPPER that uses container_of.   Part of why this looks so weird is that 'shrinkfn' name, it's not a function its the shrinker struct.  Renaming that might make it all a bit clearer.